<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Home · PlmDCA</title><script data-outdated-warner src="assets/warner.js"></script><link rel="canonical" href="https://pagnani.github.io/PlmDCA.jl/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href>PlmDCA</a></span></div><form class="docs-search" action="search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Home</a><ul class="internal"><li><a class="tocitem" href="#Package-Features"><span>Package Features</span></a></li><li><a class="tocitem" href="#Overview"><span>Overview</span></a></li><li><a class="tocitem" href="#Usage"><span>Usage</span></a></li><li><a class="tocitem" href="#Multithreading"><span>Multithreading</span></a></li><li><a class="tocitem" href="#Load-PlmDCA-package"><span>Load PlmDCA package</span></a></li><li><a class="tocitem" href="#Learn-the-parameters"><span>Learn the parameters</span></a></li><li><a class="tocitem" href="#Contact-Prediction"><span>Contact Prediction</span></a></li><li><a class="tocitem" href="#index"><span>Methods Reference</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Home</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Home</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/pagnani/PlmDCA.jl/blob/master/docs/src/index.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="PlmDCA"><a class="docs-heading-anchor" href="#PlmDCA">PlmDCA</a><a id="PlmDCA-1"></a><a class="docs-heading-anchor-permalink" href="#PlmDCA" title="Permalink"></a></h1><ul><li>Pseudo-likelihood maximization for protein.</li></ul><h2 id="Package-Features"><a class="docs-heading-anchor" href="#Package-Features">Package Features</a><a id="Package-Features-1"></a><a class="docs-heading-anchor-permalink" href="#Package-Features" title="Permalink"></a></h2><ul><li>Learn pseudo-likelihood model from multiple sequence alignment</li></ul><p>See the <a href="#index">Index</a> for the complete list of documented functions and types.</p><h2 id="Overview"><a class="docs-heading-anchor" href="#Overview">Overview</a><a id="Overview-1"></a><a class="docs-heading-anchor-permalink" href="#Overview" title="Permalink"></a></h2><p>Protein families are given in form of multiple sequence alignments (MSA) <span>$D = (a^m_i |i = 1,\dots,L;\,m = 1,\dots,M)$</span> of <span>$M$</span> proteins of aligned length <span>$L$</span>. The entries <span>$a^m_i$</span> equal either one of the standard 20 amino acids, or the alignment gap <span>$–$</span>. In total, we have <span>$q = 21$</span> possible different symbols in D. The aim of unsupervised generative modeling is to earn a statistical model <span>$P(a_1,\dots,a_L)$</span> of (aligned) full-length sequences, which faithfully reflects the variability found in <span>$D$</span>: sequences belonging to the protein family of interest should have comparably high probabilities, unrelated sequences very small probabilities. Here we propose a computationally efficient approach based on pseudo-likelihood maximization. </p><p>We start from the exact decomposition <span>$P_i(a_i| a_{\setminus i})$</span> where <span>$a_{\setminus i} := {a_1,\dots,a_{i-1},a_{i+1},\dots,a_i$</span>, i.e. all residues of the sequence of amino acids but the one relative to residue <span>$i$</span>. </p><p>Here, we use the following parametrization:</p><p class="math-container">\[P(a_i |a_{\setminus i}) = \frac{\exp \left\{ h_i(a_i) + \sum_{j\neq i}
J_{i,j}(a_i,a_j)\right\} }{z_i(a_{\setminus i})}\,,\]</p><p>where:</p><p class="math-container">\[z_i(a_{\setminus i})= \sum_{a=1}^{q} \exp \left\{ h_i(a) + \sum_{j=1\neq i} J_{i,j}(a,a_j)\right\} \,,\]</p><p>is the normalization factor. The pseudo-likelihood maximization strategy, aims at finding the value of the <span>$J, h$</span> parameters, that maximize the log-likelihood:</p><p>$ {\cal L} = \frac{1}{M} \sum<em>{m=1}^M  \left(h</em>i(a^m<em>i) + \sum</em>{j\neq i} J<em>{ij}(a</em>i^m,a<em>j^m) - \log z</em>i(a^m_{\setminus i})\right)\,.$</p><p>In machine learning, this parametrization is known as the {\em soft-max regression}, a generalization of logistic regression to multi-class labels.</p><h2 id="Usage"><a class="docs-heading-anchor" href="#Usage">Usage</a><a id="Usage-1"></a><a class="docs-heading-anchor-permalink" href="#Usage" title="Permalink"></a></h2><p>The typical pipeline to use the package is:</p><ul><li>Compute PlmDCA parameters from a multiple sequence alignment:</li></ul><pre><code class="nohighlight hljs">julia&gt; res=plmdca(filefasta; kwds...)</code></pre><h2 id="Multithreading"><a class="docs-heading-anchor" href="#Multithreading">Multithreading</a><a id="Multithreading-1"></a><a class="docs-heading-anchor-permalink" href="#Multithreading" title="Permalink"></a></h2><p>To fully exploit the the multicore parallel computation, julia should be invoked with</p><pre><code class="nohighlight hljs">$ julia -t nthreads # put here nthreads equal to the number of cores you want to use</code></pre><p>If you want to set permanently the number of threads to the desired value, you can either create a default environment variable <code>export JULIA_NUM_THREADS=24</code> in your <code>.bashrc</code>. More information <a href="https://docs.julialang.org/en/v1.6/manual/multi-threading/">here</a></p><h2 id="Load-PlmDCA-package"><a class="docs-heading-anchor" href="#Load-PlmDCA-package">Load PlmDCA package</a><a id="Load-PlmDCA-package-1"></a><a class="docs-heading-anchor-permalink" href="#Load-PlmDCA-package" title="Permalink"></a></h2><p>The package is on the General Registry. It can be installed from the package manager by</p><pre><code class="nohighlight hljs">pkg&gt; add PlmDCA</code></pre><p>and </p><pre><code class="nohighlight hljs">julia&gt; using PlmDCA</code></pre><h2 id="Learn-the-parameters"><a class="docs-heading-anchor" href="#Learn-the-parameters">Learn the parameters</a><a id="Learn-the-parameters-1"></a><a class="docs-heading-anchor-permalink" href="#Learn-the-parameters" title="Permalink"></a></h2><p>There are two different learning strategies: </p><ul><li>The asymmetric one invoked by the <code>plmdca_asym</code> method (the <code>plmdca</code> method points to the asymmetric strategy)</li><li>The symmetric strategy, invoked by the <code>plmdca_sym</code>. This method is slower and typically less accurate.</li></ul><p>Both methods return the parameters  <code>res::PlmOut</code>, a <code>struct</code> containing:</p><ol><li><code>Jtensor</code>: the 4 dimensional Array (q×q×L×L)  of the couplings in zero-sum gauge J[a,b,i,j]</li><li><code>Htensor</code>: the 2 dimensional Array (q×L) of the fields in zero-sum gauge h[a,i]</li><li><code>pslike</code>: a vector of length <code>L</code> containing the log-pseudolikelihoods</li><li><code>score</code>: a vector of `<code>(i,j,val)::Tuple{Int,Int,Float64}</code> containing the DCA score <code>val</code> relative to the <code>i,j</code> pair in descending order.</li></ol><p>For both methods, the keyword arguments (with their default value) are:</p><ul><li><code>epsconv::Real=1.0e-5</code> (convergence parameter)</li><li><code>maxit::Int=1000</code> (maximum number of iteration - don&#39;t change)</li><li><code>verbose::Bool=true</code> (set to <code>false</code> to suppress printing on screen)</li><li><code>method::Symbol=:LD_LBFGS</code> (optimization method)</li></ul><pre><code class="nohighlight hljs">res=plmdca(&quot;data/PF14/PF00014_mgap6.fasta.gz&quot;, verbose=false, lambdaJ=0.02,lambdaH=0.001);</code></pre><h2 id="Contact-Prediction"><a class="docs-heading-anchor" href="#Contact-Prediction">Contact Prediction</a><a id="Contact-Prediction-1"></a><a class="docs-heading-anchor-permalink" href="#Contact-Prediction" title="Permalink"></a></h2><p>Contact prediction is contained in the the output of <code>plmdca</code>. <code>::Output</code> contains a <code>score</code> field which is a  <code>Vector</code> of <code>Tuple</code> ranked in descending score order. Each <code>Tuple</code> contains <span>$i,j,s_{ij}$</span> where <span>$s_{ij}$</span> is the DCA score of the residue pair <span>$i,j$</span>. The residue numbering is that of the MSA, and not of the unaligned full sequences.</p><h2 id="index"><a class="docs-heading-anchor" href="#index">Methods Reference</a><a id="index-1"></a><a class="docs-heading-anchor-permalink" href="#index" title="Permalink"></a></h2><ul><li><a href="#PlmDCA.plmdca-Tuple{String}"><code>PlmDCA.plmdca</code></a></li><li><a href="#PlmDCA.plmdca_asym-Tuple{String}"><code>PlmDCA.plmdca_asym</code></a></li><li><a href="#PlmDCA.plmdca_asym-Union{Tuple{T}, Tuple{Matrix{T}, Vector{Float64}}} where T&lt;:Integer"><code>PlmDCA.plmdca_asym</code></a></li></ul><article class="docstring"><header><a class="docstring-binding" id="PlmDCA.plmdca-Tuple{String}" href="#PlmDCA.plmdca-Tuple{String}"><code>PlmDCA.plmdca</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">plmdca(filename::String; kwds...) -&gt; PlmOut</code></pre><p>Run <a href="#PlmDCA.plmdca_asym-Tuple{String}"><code>plmdca_asym</code></a> on the fasta alignment in <code>filename</code> It returns a struct <code>::PlmOut</code> containing four fields:</p><ol><li><p><code>Jtensor</code>: the 4 dimensional Array (q×q×L×L)  of the couplings in zero-sum gauge J[a,b,i,j]</p></li><li><p><code>Htensor</code>: the 2 dimensional Array (q×L) of the fields in zero-sum gauge h[a,i]</p></li><li><p><code>pslike</code>: a vector of length <code>L</code> containining the log-pseudolikelihoods</p></li><li><p><code>score</code>: a vector of `<code>(i,j,val)::Tuple{Int,Int,Float64}</code> containing the DCA score <code>val</code> relative to the <code>i,j</code> pair in descending order.</p><p>Optional arguments:</p></li></ol><ul><li><code>lambdaJ::Real=0.01</code> coupling L₂ regularization parameter (lagrange multiplier)</li><li><code>lambdaH::Real=0.01</code> field L₂ regularization parameter (lagrange multiplier)</li><li><code>epsconv::Real=1.0e-5</code> convergence value in minimzation</li><li><code>maxit::Int=1000</code> maximum number of iteration in minimization</li><li><code>verbose::Bool=true</code> set to <code>false</code> to stop printing convergence info on <code>stdout</code></li><li><code>method::Symbol=:LD_LBFGS</code> optimization strategy see <a href="https://github.com/JuliaOpt/NLopt.jl"><code>NLopt.jl</code></a> for other options</li></ul><p><strong>Example</strong></p><pre><code class="nohighlight hljs">julia&gt; res = plmdca_asym(&quot;data/pf00014short.fasta&quot;,lambdaJ=0.01,lambdaH=0.01,epsconv=1e-12);</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/pagnani/PlmDCA.jl/blob/8adbfedefb4e25f28476c82a29f9b184f939154b/src/plmdca_asym.jl#L89-L113">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PlmDCA.plmdca_asym-Tuple{String}" href="#PlmDCA.plmdca_asym-Tuple{String}"><code>PlmDCA.plmdca_asym</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">plmdca_asym(filename::String; kwds...)-&gt; PlmOut</code></pre><p>Run <a href="#PlmDCA.plmdca_asym-Tuple{String}"><code>plmdca_asym</code></a> on the fasta alignment in <code>filename</code> It returns a struct <code>::PlmOut</code> containing four fields:</p><ol><li><p><code>Jtensor</code>: the 4 dimensional Array (q×q×L×L)  of the couplings in zero-sum gauge J[a,b,i,j]</p></li><li><p><code>Htensor</code>: the 2 dimensional Array (q×L) of the fields in zero-sum gauge h[a,i]</p></li><li><p><code>pslike</code>: a vector of length <code>L</code> containining the log-pseudolikelihoods</p></li><li><p><code>score</code>: a vector of `<code>(i,j,val)::Tuple{Int,Int,Float64}</code> containing the DCA score <code>val</code> relative to the <code>i,j</code> pair in descending order.</p><p>Optional arguments:</p></li></ol><ul><li><code>lambdaJ::Real=0.01</code> coupling L₂ regularization parameter (lagrange multiplier)</li><li><code>lambdaH::Real=0.01</code> field L₂ regularization parameter (lagrange multiplier)</li><li><code>epsconv::Real=1.0e-5</code> convergence value in minimzation</li><li><code>maxit::Int=1000</code> maximum number of iteration in minimization</li><li><code>verbose::Bool=true</code> set to <code>false</code> to stop printing convergence info on <code>stdout</code></li><li><code>method::Symbol=:LD_LBFGS</code> optimization strategy see <a href="https://github.com/JuliaOpt/NLopt.jl"><code>NLopt.jl</code></a> for other options</li></ul><p><strong>Example</strong></p><pre><code class="nohighlight hljs">julia&gt; res = plmdca_asym(&quot;data/pf00014short.fasta&quot;,lambdaJ=0.01,lambdaH=0.01,epsconv=1e-12);</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/pagnani/PlmDCA.jl/blob/8adbfedefb4e25f28476c82a29f9b184f939154b/src/plmdca_asym.jl#L54-L78">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PlmDCA.plmdca_asym-Union{Tuple{T}, Tuple{Matrix{T}, Vector{Float64}}} where T&lt;:Integer" href="#PlmDCA.plmdca_asym-Union{Tuple{T}, Tuple{Matrix{T}, Vector{Float64}}} where T&lt;:Integer"><code>PlmDCA.plmdca_asym</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">plmdca_asym(Z::Array{T,2},W::Vector{Float64}; kwds...) -&gt; ::PlmOut</code></pre><p>Pseudolikelihood maximization on the M×N alignment <code>Z</code> (numerically encoded in  1,…,21), and the <code>M</code>-dimensional normalized weight vector <code>W</code>.</p><p>It returns a struct <code>::PlmOut</code> containing four fields:</p><ol><li><p><code>Jtensor</code>: the 4 dimensional Array (q×q×L×L)  of the couplings in zero-sum gauge J[a,b,i,j]</p></li><li><p><code>Htensor</code>: the 2 dimensional Array (q×L) of the fields in zero-sum gauge h[a,i]</p></li><li><p><code>pslike</code>: a vector of length <code>L</code> containining the log-pseudolikelihoods</p></li><li><p><code>score</code>: a vector of `<code>(i,j,val)::Tuple{Int,Int,Float64}</code> containing the DCA score <code>val</code> relative to the <code>i,j</code> pair in descending order.</p><p>Optional arguments:</p></li></ol><ul><li><code>lambdaJ::Real=0.01</code> coupling L₂ regularization parameter (lagrange multiplier)</li><li><code>lambdaH::Real=0.01</code> field L₂ regularization parameter (lagrange multiplier)</li><li><code>epsconv::Real=1.0e-5</code> convergence value in minimzation</li><li><code>maxit::Int=1000</code> maximum number of iteration in minimization</li><li><code>verbose::Bool=true</code> set to <code>false</code> to stop printing convergence info on <code>stdout</code></li><li><code>method::Symbol=:LD_LBFGS</code> optimization strategy see <a href="https://github.com/JuliaOpt/NLopt.jl"><code>NLopt.jl</code></a> for other options</li></ul><p><strong>Example</strong></p><pre><code class="nohighlight hljs">julia&gt; res = plmDCA(Z,W,lambdaJ=0.01,lambdaH=0.01,epsconv=1e-12);</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/pagnani/PlmDCA.jl/blob/8adbfedefb4e25f28476c82a29f9b184f939154b/src/plmdca_asym.jl#L1-L27">source</a></section></article></article><nav class="docs-footer"><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.24 on <span class="colophon-date" title="Monday 29 May 2023 12:02">Monday 29 May 2023</span>. Using Julia version 1.9.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
